File: language/src/main/java/brj/nodes/IfNode.java
package brj.nodes;

import brj.BridjeLanguage;
import brj.BridjeTypesGen;
import com.oracle.truffle.api.frame.VirtualFrame;
import com.oracle.truffle.api.nodes.UnexpectedResultException;
import com.oracle.truffle.api.profiles.ConditionProfile;
import com.oracle.truffle.api.source.SourceSection;
import org.jetbrains.annotations.NotNull;

import static com.oracle.truffle.api.CompilerDirectives.shouldNotReachHere;

public class IfNode extends ExprNode {

    @Child
    private ExprNode predNode, thenNode, elseNode;

    private final SourceSection sourceSection;

    private final ConditionProfile profile = ConditionProfile.createBinaryProfile();

    public IfNode(BridjeLanguage lang, ExprNode predNode, ExprNode thenNode, ExprNode elseNode, SourceSection sourceSection) {
        super(lang);
        this.predNode = predNode;
        this.thenNode = thenNode;
        this.elseNode = elseNode;
        this.sourceSection = sourceSection;
    }

    @Override
    public SourceSection getSourceSection() {
        return sourceSection;
    }

    @NotNull
    @Override
    public Object execute(@NotNull VirtualFrame frame) {
        boolean branch;
        try {
            branch = profile.profile(BridjeTypesGen.expectBoolean(predNode.execute(frame)));
        } catch (UnexpectedResultException e) {
            throw shouldNotReachHere(e);
        }

        return branch ? thenNode.execute(frame) : elseNode.execute(frame);
    }
}

File: language/src/main/kotlin/brj/nodes/IfNode.kt
package brj.nodes

import brj.BridjeNode
import com.oracle.truffle.api.frame.VirtualFrame
import com.oracle.truffle.api.profiles.CountingConditionProfile

class IfNode(
    @field:Child private var predNode: BridjeNode,
    @field:Child private var thenNode: BridjeNode,
    @field:Child private var elseNode: BridjeNode
) : BridjeNode() {

    private val conditionProfile = CountingConditionProfile.create()

    override fun execute(frame: VirtualFrame): Any? {
        val condition = predNode.executeBoolean(frame)
        return if (conditionProfile.profile(condition)) {
            thenNode.execute(frame)
        } else {
            elseNode.execute(frame)
        }
    }
}
